[Blackjack - Step1] Submit san-ghun's mission

san-ghun opened the pull request:
# Blackjack Step 1

## Overview

This PR is to implement requirements of the Blackjack project step1. The project is a simulation of card game Blackjack in terminal text input and output. The step1 requirements are for a simplified version of Blackjack. In this game, the player or dealer with a total closest to 21 without going over wins.

## Main Features

- Implement and design under application of MVC pattern.
   - Models: [ Card, Dealer, Hand, Player, Rank, Stats ] + [ Playable (interface) ]
   - Views: [ InputView, OutputView ]
   - Controllers: [ GameManager, CardManager, PlayerManager, DealerManager, StatsManager ]
   - Utils: [ CardGenerator ]

- `GameManager` is an object and a main controller unit among the controllers to run program and other managers working under `GameManager`.
- `CardManager` take one input parameter `CardGenerator` which can be extracted into inheritable object in future and can be used as DI(Dependency Injection).
- interface `Playable` inherit to `Player` and `Dealer` to share role as playable in game.
- `Stats` calculates and stores final results of players and dealer.

## Additional Notes

In this project, I tried to apply MVC pattern under my understanding and it was mostly focused on how to separate responsibility of a main controller. That is why I have several managers as controller units. Also wanted to follow class design to limit access to other class and object. TDD is still hard behavior to apply during the project but my brain is getting used to it. I am really glad that I am getting better than previous project in many aspect.

Hope I get a lot of feedback from your review, so I can learn and grow further more.

Thank you!

san-ghun committed:
docs(readme): add Features List

- add features for CardGenerator in new section Features List

san-ghun committed:
feat(Card,CardGenerator): add a value class and methods to generate a card

- add a value class 'Card' to hold value of suit
- add generateSuits() method which create a list of combinations for suits
- add generateCard() method which create a card instance
- add tests for new methods

san-ghun committed:
feat(CardGenerator): add 'generateCards()' method to create a list of cards

- add two test cases for the methods

san-ghun committed:
feat(InputView): add 'readPlayerNames' method and 'readYesOrNo' method

- add readPlayerNames to take user input for players names
- add readYesOrNo to take user input for 'yes or no' drawing another card

san-ghun committed:
feat(Card,Rank): update 'Card' and add 'Rank'

- change 'Card' class from value class to data class for the usage of multiple properties
- add 'Rank' enum class for later usage of calculating score of hand, which is a collection of cards
- add 'CardTest' and add a test case to validate generation of a property 'digit'

san-ghun committed:
feat(CardManager): add CardManager class with test and update README.md file

- add new class 'CardManager' as a controller unit to manage cards
- add 'checkCardMap()' method to figure out whether the card is available to take or not
- add 'giveCard()' method to get card from cards and update cardmap to check availability and return the card
- add test cases for methods in the class
- update Features List for CardManager under Controller section

san-ghun committed:
feat(Playable): add new interface 'Playable' for a character of player and dealer

san-ghun committed:
feat(Player,Rank): add data class 'Player' and update enum class 'Rank'

- data class 'Player' inherit interface 'Playable'
- add test for 'Player' in 'PlayerTest'
- delete 'BIG_ACE' from 'Rank'
- update features list in README.md file

san-ghun committed:
feat(Dealer): add data class 'Dealer'

- data class 'Dealer' inherit interface 'Playable'
- add test for 'Dealer' in 'DealerTest'
- update features list in README.md file
- remove a line that not used anymore from `PlayerTest`

san-ghun committed:
feat(GameManager): add new class 'GameManager' as a controller unit

- Working In Progress
- write rough pseudo class with comments to implement 'GameManager'
- add new section 'GameManager' in features list, README.md file

san-ghun committed:
feat(OutputView,GameManager,Application,Models): add OutputView and apply to GameManager

- add `OutputView` with methods to display the game data to user
- apply the change to `GameManager`
- add `getStringOfHand()` to `Player` and `Dealer` to make smoother interaction with `OutputView`
- add `main()` method is `Application.kt` file to run program
- update feature list in README

san-ghun committed:
feat(Stats,OutputView,GameManager): add data class 'Stats' to manage final result of game

- add 'initPlayerBoard()' to collect result of players against dealer
- add 'updateDealerStats()' to calculate and collect total score of dealer
- add 'displayFinalResult()' in 'OutputView' to display the final result based on 'Stats'
- apply these changes into 'GameManager'
- update features list in README.md file

san-ghun committed:
refactor(GameManager): tidy flow control of 'askPlayerHit()' less messy

san-ghun committed:
chore(Stats): make 'initPlayerBoard()' private

san-ghun committed:
test(StatsTest): add test for 'Stats' class

san-ghun committed:
refactor(Hand,Models,OutputView,PlayerTest): extract 'hand: List<Card>' property and logic into dedicated class 'Hand'

- extract 'hand' property and logic related with it into new data class 'Hand'
- move repeated implementation in 'Dealer' and 'Player' into interface 'Playable'
- modify 'Player', 'Dealer', 'OutputView', and 'PlayerTest' derived by the extraction of new class 'Hand'

san-ghun committed:
refactor(Stats): separate condition part from 'initPlayerBoard()'

- add new private method 'recordPlayerBoard()'

san-ghun committed:
refactor(Card,Rank,Hand): simplify 'of' method in companion object of 'Rank'

- let 'of' method in 'Rank' dynamically find and return dedicated 'Rank'
- add new property 'value' in 'Card' data class to store value of 'Rank' of 'Card'
- refactor and simplify 'calculateCards()' method in 'Hand' data class

san-ghun committed:
refactor(Stats): add constants to improve readability of game result

- use companion object to assign constant values for win/lose/tie

san-ghun committed:
refactor(OutputView): add new method for refactoring and add constants to improve readability of player results

- extract conditional part of 'displayFinalResults()' into new private method 'givePlayerResult()'
- use private object to assign constant values for win/lose/tie

san-ghun committed:
fix(InputView): add input validation for blank input with 'require()'

- add TODO comment in GameManager for later

san-ghun committed:
test(HandTest): add test for 'Hand' data class

san-ghun committed:
feat(Controller): add 'PlayerManager' as a controller unit and apply to 'GameManager'

- extract 'Player' related parts from 'GameManager' to new class 'PlayerManager'

san-ghun committed:
feat(Controller): add 'DealerManager' as a controller unit and apply to 'GameManager'

san-ghun committed:
feat(Controller): add 'StatsManager' as a controller unit and apply to 'GameManager'

- extract statistics related parts from 'GameManager' to 'StatsManager'

san-ghun committed:
feat(InputView): add new method 'retryable()' to re-prompt user input for invalid input

san-ghun committed:
refactor(Controllers,Application): extract view parts from 'PlayerManager' and 'StatsManager' to 'GameManager'

- change 'GameManger' from class to object to make it singleton unit
- split 'run()' method into parts to separate responsibilities into parts, in 'GameManager'
- add 'InputView' and 'OutputView' as properties in 'GameManger' to have access to views
- in 'PlayerManager' add two methods to call 'GameManager' and access to view methods

san-ghun committed:
refactor(OutputView): update 'displayInitialState()' method

- change string formation process for 'displayInitialState()' method
- remove some comments in 'Dealer'

san-ghun committed:
refactor(PlayerManager,StatsManager): cover open properties and methods into private

san-ghun committed:
fix(InputView): add input validation for duplicates in names for 'readPlayerNames()'

san-ghun committed:
docs(readme): mark and update features list in README.md file

san-ghun committed:
chore: remove resolved TODO comment

Wordbe commented on src/main/kotlin/blackjack/Application.kt:
```diff
@@ -0,0 +1,8 @@
+import blackjack.controller.GameManager
+
+fun main() {
+    val gm = GameManager
```
I think it does not need a variable!

san-ghun replied:
Agree! Thanks for pointing out!

Wordbe commented on src/main/kotlin/blackjack/controller/CardManager.kt:
```diff
@@ -0,0 +1,46 @@
+import blackjack.model.Card
+import blackjack.utils.CardGenerator
+
+class CardManager(generator: CardGenerator = CardGenerator) {
```
Object instances usually don’t need to receive dependencies through their constructors. It’s good to think about why that is! 🤔

san-ghun replied:
Thank you for a topic to consider. 
What I can answer is, in this case, I guess receiving one-use only dependency(`CardGenerator`) is an overkill(?) for the `CardManager` instance.

san-ghun replied:
I am considering to redesign `Card` class to have feature of generating `Cards` that only hold 52 instances of `Card` and only return pointer of the instance, preventing to create new `Card` instance any occasion.

Wordbe replied:
Great Thank you!

Wordbe commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -0,0 +1,6 @@
+package blackjack.model
+
+data class Card(val name: String) {
+    val digit = name.substring(0, name.length - 1)
```
Parsing strings can easily cause issues if the format changes, and it’s also hard for people to read. Instead, try creating objects with a defined structure!

san-ghun replied:
Yes, I agree. I was also skeptical to assign parsed string into variable. I reimplement "Card" class with two enum class, I hope resubmitted implementation is better with you review.

Wordbe commented on src/main/kotlin/blackjack/controller/GameManager.kt:
```diff
@@ -0,0 +1,51 @@
+import blackjack.view.InputView
+import blackjack.view.OutputView
+
+object GameManager {
```
I think using `~Manager` might be overkill. Especially if you just access something like dealManager.dealer directly, it defeats the purpose of having a Manager! 🙂

san-ghun replied:
Yes, I agree on lack of efficiency on 'DealerManager'.
But is it okay to separate controllers into several to controller dedicated models? It is my first try separating a main controller into many. I am not sure, is it a good approach to separate responsibility of the main controller?

Wordbe replied:
As I mentioned, the current structure feels a bit excessive.
Also, rather than naming everything as "Manager", it would be better to give names that reflect the domain's role—so they act as living objects with clear responsibilities. 🙂

Wordbe commented on src/main/kotlin/blackjack/controller/GameManager.kt:
```diff
@@ -0,0 +1,51 @@
+    val playerManager = PlayerManager()
+    val dealerManager = DealerManager()
+    val statsManager = StatsManager(playerManager.players, dealerManager.dealer)
+    val inputView = InputView
```
There’s no need to assign an object to a variable. Especially since some parts use this variable while others call static methods, mixing the two ways.

san-ghun replied:
Thanks! I also thought about the needs for assigning object into variable. It is good to know from your review.

Wordbe commented on src/main/kotlin/blackjack/controller/PlayerManager.kt:
```diff
@@ -0,0 +1,48 @@
+import blackjack.model.Player
+
+class PlayerManager {
+    private var _players: List<Player> = emptyList()
```
Why did you use backing property? What benefits does it have? 🙂

san-ghun replied:
I used backing property because I wanted to not expose `players` to be editable. I want to expose `players` read-only and only able to add or remove by `PlayerManager`.

Wordbe replied:
Great! But since it may be hard to read, please use it only where it's truly necessary. 😊

Wordbe commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,22 @@
+package blackjack.model
+
+data class Hand(val cards: List<Card> = emptyList()) {
```
Why did you use data class? 🙂

san-ghun replied:
Well, I though I am going to override some data class method to manipulate `Hand` for output view with `toString()` or addition method. But eventually it didn't go the way I have thought. I am making it into normal class.

Wordbe replied:
Good Thank you!

Wordbe commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,22 @@
+package blackjack.model
+
+data class Hand(val cards: List<Card> = emptyList()) {
```
Let's define the numbers in this class as meaningful constants to improve readability.

san-ghun replied:
Yes! Thanks for letting me know gracefully!

Wordbe commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,22 @@
+
+    fun shouldDrawCardOrNot(): Boolean {
+        val score = calculateHand()
+        return score <= 16
```
Shall we define the magic numbers as constants?

san-ghun replied:
Yes! I am on it! Thanks!!

Wordbe commented on src/main/kotlin/blackjack/model/Rank.kt:
```diff
@@ -0,0 +1,25 @@
+
+    companion object {
+        fun of(digit: String): Rank {
+            return Rank.entries.firstOrNull { it.digit == digit } ?: NONE
```
How about throwing an exception instead of using NONE?

san-ghun replied:
Yes! Good point! It is so much better!

Wordbe commented on src/main/kotlin/blackjack/model/Stats.kt:
```diff
@@ -0,0 +1,48 @@
+package blackjack.model
+
+data class Stats(val players: List<Player>, val dealer: Dealer) {
+    private var _playerBoard = initPlayerBoard()
```
Instead of defining internal variables, try returning the result from a method.
Going forward, when creating state in an object, do so only when it's truly necessary. 💡

san-ghun replied:
Yes, now I know what is difference between calculated property and instance property(or field). 
I remove backing property for `playerBoard` and replace it into calculated property.
Thanks for pointing out!

Wordbe commented on src/main/kotlin/blackjack/model/Stats.kt:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    companion object {
+        const val LOSE = 0
```
Try defining it using enums or descriptive strings instead of raw numbers. It will improve the readability!

san-ghun replied:
Yes, agree! I am introducing enums for this and in `OutputView` object. Thank you!

Wordbe commented on src/main/kotlin/blackjack/utils/CardGenerator.kt:
```diff
@@ -0,0 +1,33 @@
+        return values.toList()
+    }
+
+    fun generateCard(suit: String): Card {
```
This seems unnecessary. Try removing it and simplifying the code. 😊

san-ghun replied:
Yes, I removed the `CardGenerator` object including the methods.

Wordbe reviewed (changes requested):
Hello, San!
Nice to meet you! 😊

Thank you for your hard work writing the Blackjack mission code with care.
Please review the feedback and apply any necessary changes after considering them!

Here are some general areas that stood out for improvement:

Use backing properties only when truly necessary. They can make the code harder to read and understand. Instead of using numbers to represent states, consider using enums. They make the code much more readable. Avoid creating and parsing strings unnecessarily. Manage data in structured formats from the beginning—it’ll be easier to maintain. Strive to simplify your code as much as possible. Simple code is always easier to manage than complex logic. Replace magic numbers with constants. This improves clarity, reduces mistakes, and enhances overall understanding. 

Thank you again! Keep up the great work! 🙌

san-ghun committed:
docs(readme): add section for reminders to keep in mind

- add new section 'Reminders for Improvement from Review from --wordbe'

san-ghun committed:
refactor(Application): remove unneeded value assignment in 'main()'

- call directly to 'GameManager' to 'run()'

san-ghun committed:
refactor(controller/*,model/*,OutputView): reimplement 'Card' class into 'PlayingCard' class

Problem
- There are high possibility of creating for 'Card' instances those holding same values, and might provide confusion of managing 'Card' instances.

Solution
- redesign card class, removing 'Card' class and adding new class 'PlayingCard' with companion object that holding only 52 instances of 'PlayingCard'
- move card generation logic from 'CardGenerator' object into 'PlayingCard' class
- remove 'CardGenerator' object
- add 'Suit' enum class to create 'PlayingCard' instance using with 'Rank' enum class

Others
- apply new added class 'PlayingCard' in all card related lines of code

Test
- add new object 'Fixture' to provide easy fixture of instance to test code

san-ghun committed:
refactor(DealerManager,GameManager): remove 'DealerManager' class, give 'GameManager' direct access to 'dealer'

san-ghun committed:
refactor(GameManager,PlayerManager): remove variable assignment for views

- change object name 'GameManager' to 'GameMaster' to make difference between object and class
- remove variable assignment for views in 'GameMaster'
- give 'PlayerManager' direct access to views and remove related dangling methods

san-ghun committed:
refactor(Hand): alter `Hand` from data class to class

- at the beginning of design for 'Hand' was to manipulate the instance to handle addition and string output with data class methods, but it didn't go that way, and forgot to recall

san-ghun committed:
refactor(Playable,Dealer,Hand): alter magic number into constants

san-ghun committed:
refactor(Result,Stats,OutputView): introduce 'Result' enum class to represent result of players

- extract lose/win/tie notation from 'Stats' and 'OutputView' introducing new enum class 'Result'

san-ghun committed:
refactor(StatsManager,GameMaster): delete 'StatsManager' for lack of usability

- remove logic in 'StatsManager' and move the logic into 'GameMaster'

san-ghun commented:
Hey, Suno!

Thanks for your thoughtful review! I have managed to apply review you mentioned. A main change since your last review is that I have reimplemented `Card` class into `PlayingCard` class with enum classes `Rank` and `Suit`. Because of the reimplementation of "PlayingCard" class, I edited lines of code. Sorry for adding confusion by reimplementing of class and hope you understand. 

# Main Changes

## Replacement of `Card` class to `PlayingCard` class

**Problem**: There are high possibility of creating for 'Card' instances those holding same values, and might provide confusion of managing 'Card' instances.

**Solution**
- redesign card class, removing 'Card' class and adding new class 'PlayingCard' with companion object that holding only 52 instances of 'PlayingCard'
- move card generation logic from 'CardGenerator' object into 'PlayingCard' class
- remove 'CardGenerator' object
- add 'Suit' enum class to create 'PlayingCard' instance using with 'Rank' enum class

## Remove `DealerManager` class

**Problem**: 'DealerManager' has no functionality over direct access to 'Dealer' instance.

**Solution**
- remove 'DealerManager' and give 'GameMaster' a direct access to 'Dealer' instance

## Introduction of `Result` enum class to represent 'result' of game

**Problem**: Previously 'Stats' class and 'OutputView' used number to represent result of game for players and dealer. It is having duplication of same code and usage of numbers to represent result of game provide lack of readability.

**Solution**
- introduce new enum class 'Result' to represent the result of game for players and dealer

## Remove magic numbers

**Problem**: There are too many magic numbers in codebase.

**Solution**
- introduce constants for magic numbers inside of `companion object`

---

I have additional questions related with the code review using GitHub.

1. When can or should "resolve" 'conversation' or 'feedback' or 'review comment'? 
    - Should I let it remain unresolved, or can I resolve after I applying feedback in new commit?
2. What would be your recommendation to have a good manner of code review, as reviewee? 
    - I am not sure leaving comments of message like what I did is good or bad.
3. What does "reviewer" expect from "reviewee"? 
    - I think with this answer, I can provide more and precise context to my future reviews.

Thanks you!

Wordbe commented on src/main/kotlin/blackjack/controller/CardManager.kt:
```diff
@@ -0,0 +1,47 @@
+        _cardMap = initiateCardMap()
+    }
+
+    private fun initiateCardMap(): Map<PlayingCard, Boolean> {
```
Why is a Map needed here? What does the Boolean in the Map represent?
How could this be made clearer?
Is it possible to implement this without using a Map?

san-ghun replied:
Yes, thanks for pointing out! I also cannot find any use of it. I will remove the Map.
I am also thinking to remove `CardManager` and introduce new model class like `Deck` for the functionality.

Wordbe replied:
Great. It seems a nice object. Thank you.

Wordbe commented on src/main/kotlin/blackjack/controller/PlayerManager.kt:
```diff
@@ -0,0 +1,42 @@
+        _players = mutableList.toList()
+    }
+
+    // TODO: resolve indent depth
```
Remove the TODO and simplify the code as much as possible! ✂️

san-ghun replied:
Yes, I was looking away this part because it was difficult for me to improve this part. I try my best to simplify the code as much as possible! Thanks!

Wordbe commented on src/main/kotlin/blackjack/model/Hand.kt:
```diff
@@ -0,0 +1,28 @@
+        return score
+    }
+
+    fun toText(): String {
```
If it's something needed in the View layer, how about handling it in the View instead of the Domain layer? 🙂

san-ghun replied:
Yes, you're right. I will move it to the View

Wordbe commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,26 @@
+package blackjack.model
+
+data class Dealer(override val name: String = "Dealer") : Playable {
+    private var _hand = Hand()
```
Why is a backing property necessary here? 🤔

san-ghun replied:
I wanted to private the 'hand', but I think it doesn't have to, like you mentioned. I am removing it and alter to variable.

Wordbe commented on src/main/kotlin/blackjack/model/PlayingCard.kt:
```diff
@@ -0,0 +1,28 @@
+
+class PlayingCard private constructor(val rank: Rank, val suit: Suit) {
+    val value: Int get() = rank.value
+    val string: String get() = rank.digit + suit.form
```
Also, try separating parts that are only needed in the view layer here!

Wordbe commented on src/main/kotlin/blackjack/model/Result.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack.model
+
+enum class Result(val value: Int) {
```
Why is value needed here? Could it possibly be removed? 🤔

san-ghun replied:
Yeah. Actually it has no use. lol
I am removing it.

Wordbe commented on src/main/kotlin/blackjack/model/Stats.kt:
```diff
@@ -0,0 +1,45 @@
+package blackjack.model
+
+class Stats(val players: List<Player>, val dealer: Dealer) {
+    val playerBoard: Map<Player, Result> get() = calculatePlayerBoard()
```
Is playerBoard really necessary?
Try to minimize internal state as much as possible. 🙂

san-ghun replied:
Okay, I will apply your suggestion. 
To remove `playerBoard` from `Stats`, I introduce new property 'result' to `Player` and `Dealer`. And I edit functions to calculate and update result of players and dealer.

Wordbe commented on src/test/kotlin/blackjack/Fixture.kt:
```diff
@@ -0,0 +1,20 @@
+import blackjack.model.Rank
+import blackjack.model.Suit
+
+object Fixture {
```
Good 👍

Wordbe reviewed (changes requested):
Hello!

Thank you for writing the code so carefully and for reflecting on the review.

I’ve left some comments on parts that might need further consideration, so please take a look! 😊

---

Q1. When can or should "resolve" 'conversation' or 'feedback' or 'review comment'?Should I let it remain unresolved, or can I resolve after I applying feedback in new commit?

Yes, please fix the issues, leave a comment, and then resolve the conversation! 

Q2. What would be your recommendation to have a good manner of code review, as reviewee?

If you have a reason, feel free to share your opinion politely and confidently! 🙂

Q3. I am not sure leaving comments of message like what I did is good or bad.What does "reviewer" expect from "reviewee"?I think with this answer, I can provide more and precise context to my future reviews.

It’s really helpful if you share what you worked on—what was challenging and how you addressed it. That way, we can understand your thought process better. Or you can also suggest alternative approaches if you have any! 

Thank you!

san-ghun committed:
refactor(CardManager): remove a property 'cardMap' for non-usage

san-ghun committed:
refactor(PlayerManager): simplify logic of `askPlayerHit()`

san-ghun committed:
refactor(Hand,OutputView): move 'toText()' function for 'Hand' from domain to view with new name 'handToText()'

san-ghun committed:
refactor(Playable,Dealer,Player): remove backing property 'hand' from 'Playable's

- change 'hand' property from 'val' to 'var' in 'Playable'

san-ghun committed:
refactor(PlayingCard,OutputView): move calculate property 'string' from 'PlayingCard' to 'OutputView'

san-ghun committed:
refactor(Result): remove property 'value' for no-usage

san-ghun committed:
refactor(Stats,Result,Playable,OutputView): remove property 'playerBoard' from 'Stats' and introduce property 'result' to 'Playable' to store result of player and dealer

- remove unnecessary variables in 'Stats'
- add new enum 'NONE' to 'Result' as an init value for result of Playable

san-ghun committed:
refactor(Deck,PlayingCard,GameMaster): replace 'CardManager' into new class 'Deck'

san-ghun committed:
refactor(Player,Hand,GameMaster): remove 'PlayerManager' and update 'GameMaster'

san-ghun commented:
Hey, Suno!

Thanks for your review! There are two main changes since your last review. I replaced `CardManager` into `Deck`, and removed `PlayerManager`. These changes aim to reduce complexity of controllers and make the role and names more clear and enhance readability. There are several changes to apply review you mentioned.

# Main Changes

## Replace `CardManager` and introduce new class `Deck`

**Problem**: The name 'CardManager' is too ambiguous and not clear to represent the role.

**Solution**
- introduce new class `Deck` as an wrapper collection, and remain essential functionalities for 'Deck'.

## Remove `PlayManager` and move remained logic into `GameMaster`

**Problem**: The role and name of `PlayerManager` is ambiguous, and initiation of a player is complex

**Solution**
- remove `PlayerManager`
- let players and dealer to initiate with 'hand'

# Noticeable Changes

- move View related functionalities from `Hand` and `PlayingCard` to View layer
- remove a state property `playerBoard` from `Stats` and move the feature to `Playable` with `Result`
- remove unnecessary backing properties to enhance readability
- remove unnecessary states from `GameMaster`

---

Additionally, thanks for your answers from previous comment!
Thanks for your review, all the time!

Wordbe commented on src/main/kotlin/blackjack/controller/GameMaster.kt:
```diff
@@ -0,0 +1,103 @@
+        OutputView.displayFinalResults(winStatistics)
+    }
+
+    private fun initDealer(): Dealer {
+        val dealer = Dealer()
+        dealer.hand.initCards()
+        return dealer
+    }
```
It seems more appropriate for the `Dealer` to handle this. Try moving the code inside the `Dealer`. 🙂

san-ghun replied:
Totally make sense!

Wordbe commented on src/main/kotlin/blackjack/controller/GameMaster.kt:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private fun initPlayers(): List<Player> {
+        val players = mutableListOf<Player>()
+        val names = InputView.retryable { InputView.readPlayerNames() }
+        names.forEach { name ->
+            val player = Player(name, Hand().initCards())
+            players.add(player)
+        }
+        return players.toList()
```
Instead of declaring variables beforehand, try returning a `List<Player>` in a functional style!

san-ghun replied:
Yes! I also can see how to make it better with functional style! I will try!

Wordbe commented on src/main/kotlin/blackjack/controller/GameMaster.kt:
```diff
@@ -0,0 +1,103 @@
+        }
+    }
+
+    private fun calculateStatistics(
+        players: List<Player>,
+        dealer: Dealer,
+    ): Stats {
+        val winStatistics = Stats(players, dealer)
+        winStatistics.updateDealerStats()
+        return winStatistics
+    }
```
Try assigning the role to Stats!

san-ghun replied:
I don't fully understand what you are saying but I will try on step2!

Wordbe commented on src/main/kotlin/blackjack/controller/GameMaster.kt:
```diff
@@ -0,0 +1,103 @@
+
+    private fun askPlayerHit(
+        player: Player,
+        receiveCard: () -> PlayingCard,
+    ) {
+        var isFirst = true
+
+        while (!player.isBust()) {
+            isFirst = drawOrNot(player, receiveCard, isFirst)
+            if (isFirst) break
+        }
+    }
+
+    private fun drawOrNot(
+        player: Player,
+        receiveCard: () -> PlayingCard,
+        isFirst: Boolean,
+    ): Boolean {
+        val wantsCard =
+            player.requestCard {
+                InputView.retryable { InputView.readYesOrNo(player.name) }
+            }
+
+        if (!wantsCard) {
+            if (isFirst) OutputView.displayCurrentHand(player)
+            return true
+        }
+
+        drawAndDisplayAndReturnFalse(player, receiveCard)
+        return false
+    }
+
+    private fun drawAndDisplayAndReturnFalse(
+        player: Player,
+        receiveCard: () -> PlayingCard,
+    ): Boolean {
+        player.drawCard(receiveCard())
+        OutputView.displayCurrentHand(player)
+        return false
+    }
```
The receiveCard function goes as deep as 3 levels, which makes it hard to read. It would be great to improve this.

san-ghun replied:
Yes, it is. I am going to remove 'receiveCard'.

Wordbe commented on src/main/kotlin/blackjack/model/Stats.kt:
```diff
@@ -0,0 +1,33 @@
+package blackjack.model
+
+class Stats(val players: List<Player>, val dealer: Dealer) {
+    var dealerStats = mapOf(Result.WIN to 0, Result.LOSE to 0, Result.TIE to 0)
+
+    fun updateDealerStats() {
```
It would be better to remove the state and create it within the method to return it. This could be a tradeoff between performance and state management. But if it's a lightweight calculation, I would prefer to eliminate state management. 🙂

san-ghun replied:
Okay, I may have to consider about this that with the view between performance and state management. I will keep that in mind! Thanks!

Wordbe reviewed (approved):
Hello San,

Thank you for thoughtfully considering the blackjack mission review and carefully applying the feedback.
I will approve and merge this PR here! Great work.

Please review the additional comments I left and apply them when you start the next step.
Thank you. 🙏
